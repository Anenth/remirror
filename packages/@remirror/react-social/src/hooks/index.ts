export * from './use-social';
export * from './use-social-emoji';
export * from './use-social-mention';
