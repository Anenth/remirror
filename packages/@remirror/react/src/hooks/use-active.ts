import type { AnyExtension, RemirrorManager } from '@remirror/core';

import { ReactFrameworkOutput } from '../react-types';
import { useRemirrorContext } from './use-remirror-context';

type UseActiveReturn<Extension extends AnyExtension> = ReactFrameworkOutput<
  RemirrorManager<Extension>['~E']
>['active'];

/**
 * This is a shorthand method for retrieving the active available in the
 * editor.
 *
 * ```ts
 * import { useActive } from 'remirror/react';
 * ```
 *
 * This hooks updates the local component on each state update for the editor,
 * so it can be quite expensive.
 *
 * @param autoUpdate - Set to false to prevent automatic re-rendering on every
 * state update.
 */
export function useActive<Extension extends AnyExtension = Remirror.Extensions>(
  autoUpdate = true,
): UseActiveReturn<Extension> {
  return useRemirrorContext<Extension>({ autoUpdate }).active;
}
