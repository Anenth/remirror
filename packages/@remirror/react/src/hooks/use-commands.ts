import { AnyExtension, ChainedFromExtensions, CommandsFromExtensions } from '@remirror/core';

import { useRemirrorContext } from './use-remirror-context';

/**
 * A core hook which provides the commands for usage in your editor.
 *
 * ```tsx
 * import { useCommands } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const commands = useCommands();
 *
 *   return (
 *     <>
 *       <button onClick={() => commands.toggleBold()}>
 *         Click me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 */
export function useCommands<
  Extension extends AnyExtension = Remirror.Extensions
>(): CommandsFromExtensions<Extension> {
  return useRemirrorContext<Extension>().commands;
}

/**
 * A core hook which provides the chainable commands for usage in your editor.
 *
 * ```tsx
 * import { useCommandsChain } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const chain = useCommandsChain();
 *
 *   return (
 *     <>
 *       <button onClick={() => chain.toggleBold().toggleItalic().run()}>
 *         Chain me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 */
export function useCommandsChain<
  Extension extends AnyExtension = Remirror.Extensions
>(): ChainedFromExtensions<Extension> {
  return useRemirrorContext<Extension>().chain;
}
