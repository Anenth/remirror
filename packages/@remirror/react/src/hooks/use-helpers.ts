import type { AnyExtension, RemirrorManager } from '@remirror/core';

import { ReactFrameworkOutput } from '../react-types';
import { useRemirrorContext } from './use-remirror-context';

type UseHelpersReturn<ExtensionUnion extends AnyExtension> = ReactFrameworkOutput<
  RemirrorManager<ExtensionUnion>['~E']
>['helpers'];

/**
 * A core hook which provides the helpers for usage in your editor.
 *
 * ```tsx
 * import { useHelpers } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const helpers = useHelpers();
 *
 *   return (
 *     <>
 *       <button onClick={() => Helpers.toggleBold()}>
 *         Click me!
 *       </button>
 *       <button onClick={() => Helpers.chain.toggleBold().toggleItalic().run()}>
 *         Chain me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 *
 * Passing `true` as the first argument will ensure that the component this hook
 * is placed inside will rerender on every update.
 */
export function useHelpers<ExtensionUnion extends AnyExtension = Remirror.AllExtensionUnion>(
  update = false,
): UseHelpersReturn<ExtensionUnion> {
  return useRemirrorContext<ExtensionUnion>(update ? { autoUpdate: true } : undefined).helpers;
}
