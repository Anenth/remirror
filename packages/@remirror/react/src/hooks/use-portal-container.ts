/**
 * @module
 *
 * A hook which provides access to the portal container for rendering react
 * component directly within the editor dom.
 */

import { PortalContainer } from '@remirror/extension-react-component';

import { useRemirrorContext } from './use-remirror-context';

/**
 * Retrieve the portal container which can be used to render
 */
export function usePortalContainer(): PortalContainer {
  return useRemirrorContext().portalContainer;
}
