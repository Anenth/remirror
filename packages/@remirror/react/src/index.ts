export type { UseExtensionCallback, UseRemirrorProps, UseRemirrorReturn } from './hooks';
export {
  RemirrorContext,
  useActive,
  useCommands,
  useEditorDomRef,
  useEditorState,
  useEffectWithWarning,
  useExtension,
  useForceUpdate,
  useHelpers,
  useI18n,
  useManager,
  usePortalContainer,
  useRemirror,
  useRemirrorContext,
} from './hooks';
export { createReactManager } from './react-helpers';
export type { RemirrorProps } from './react-remirror';
export { EditorComponent, Remirror } from './react-remirror';
export type {
  CreateReactManagerOptions,
  DefaultReactExtensionUnion,
  GetRootPropsConfig,
  ReactExtensionUnion,
  ReactFrameworkOutput,
  RefKeyRootProps,
  RefProps,
  UseRemirrorContextType,
} from './react-types';
export * from './renderers';
export { createEditorView } from './ssr';
