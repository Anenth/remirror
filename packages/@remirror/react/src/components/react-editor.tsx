import React, {
  cloneElement,
  Dispatch,
  ReactNode,
  Ref,
  SetStateAction,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import useEffectOnce from 'react-use/lib/useEffectOnce';
import { useFirstMountState } from 'react-use/lib/useFirstMountState';
import useUpdateEffect from 'react-use/lib/useUpdateEffect';

import {
  AnyCombinedUnion,
  bool,
  EditorWrapper,
  EditorWrapperParameter,
  EMPTY_PARAGRAPH_NODE,
  ErrorConstant,
  getDocument,
  invariant,
  isArray,
  isFunction,
  isNullOrUndefined,
  isPlainObject,
  object,
  PrimitiveSelection,
  RemirrorContentType,
  SchemaFromCombined,
  shouldUseDomEnvironment,
  UpdateStateParameter,
} from '@remirror/core';
import { EditorState } from '@remirror/pm/state';
import { ReactPreset } from '@remirror/preset-react';
import {
  addKeyToElement,
  getElementProps,
  isReactDOMElement,
  isRemirrorContextProvider,
  isRemirrorProvider,
  propIsFunction,
  RemirrorType,
} from '@remirror/react-utils';

import { usePrevious } from '../hooks';
import { PortalContainer } from '../portals';
import {
  BaseProps,
  GetRootPropsConfig,
  RefKeyRootProps,
  RemirrorContextProps,
} from '../react-types';
import { createEditorView, RemirrorSSR } from '../ssr';

/**
 * The component responsible for rendering your prosemirror editor to the DOM.
 *
 * @remarks
 *
 * This is an internal component and should only be used within the `remirror`
 * codebase. The `RemirrorProvider` is the only supported way for consuming the
 * application.
 *
 * @internal
 */
export const ReactEditor = <Combined extends AnyCombinedUnion>(
  props: ReactEditorProps<Combined>,
) => {
  const { stringHandler = defaultStringHandler, onError, manager, forceEnvironment, value } = props;

  // Cache whether this is a controlled editor.
  const isControlled = bool(value);
  const { placeholder } = props;
  const isFirstMount = useFirstMountState();

  // Update the placeholder on first render.
  if (isFirstMount && !isNullOrUndefined(placeholder)) {
    manager.getPreset(ReactPreset).setOptions({ placeholder });
  }

  // Keep the placeholder updated
  useUpdateEffect(() => {
    if (isNullOrUndefined(placeholder)) {
      return;
    }

    manager.getPreset(ReactPreset).setOptions({ placeholder });
  }, [placeholder, manager]);

  const createStateFromContent = useCallback(
    (
      content: RemirrorContentType,
      selection?: PrimitiveSelection,
    ): EditorState<SchemaFromCombined<Combined>> => {
      return manager.createState({
        content,
        doc: getDocument(forceEnvironment),
        stringHandler,
        selection,
        onError,
      });
    },
    [onError, forceEnvironment, manager, stringHandler],
  );

  const fallback = isFunction(onError) ? onError() : onError ?? EMPTY_PARAGRAPH_NODE;
  const initialContentArgs = isArray(props.initialContent)
    ? props.initialContent
    : ([props.initialContent ?? fallback] as const);
  const initialContent = initialContentArgs[0];
  const initialSelection = initialContentArgs[1];
  const initialEditorState = bool(value)
    ? value
    : createStateFromContent(initialContent, initialSelection);
  const [shouldRenderClient, setShouldRenderClient] = useState<boolean | undefined>(
    props.suppressHydrationWarning ? false : undefined,
  );

  // Store all the `logic` in a `ref`
  const methods: ReactEditorWrapper<Combined> = useEditorWrapper<Combined>({
    initialEditorState,
    setShouldRenderClient,
    createStateFromContent,
    getProps: () => props,
    getShouldRenderClient: () => shouldRenderClient,
  });

  // Handle the initial editor mount.
  useEffectOnce(() => {
    methods.onMount();

    return () => methods.onDestroy();
  });

  const previousEditable = usePrevious(props.editable);

  // Handle editor updates
  useUpdateEffect(() => {
    methods.onUpdate(previousEditable);
  }, [previousEditable]);

  // Handle controlled editor updates every time the value changes.
  useEffect(() => {
    invariant((value && isControlled) || (isNullOrUndefined(value) && !isControlled), {
      code: ErrorConstant.REACT_CONTROLLED,
      message: isControlled
        ? 'You have attempted to switch from a controlled to an uncontrolled editor. Once you set up an editor as a controlled editor it must always provide a `value` prop.'
        : 'You have provided a `value` prop to an uncontrolled editor. In order to set up your editor as controlled you must provide the `value` prop from the very first render.',
    });

    if (!value) {
      return;
    }

    methods.updateControlledState(value);
  }, [isControlled, value, methods]);

  // Return the rendered component
  return methods.render();
};

/**
 * Sets a flag to be a static remirror
 */
ReactEditor.remirrorType = RemirrorType.Editor;

/**
 * A function that takes the injected remirror params and returns JSX to render.
 *
 * @param - injected remirror params
 */
type RenderPropFunction<Combined extends AnyCombinedUnion> = (
  params: RemirrorContextProps<Combined>,
) => JSX.Element;

export interface ReactEditorProps<Combined extends AnyCombinedUnion> extends BaseProps<Combined> {
  /**
   * The render prop that takes the injected remirror params and returns an
   * element to render. The editor view is automatically attached to the DOM.
   */
  children: RenderPropFunction<Combined>;
}

class ReactEditorWrapper<Combined extends AnyCombinedUnion> extends EditorWrapper<
  Combined,
  ReactEditorProps<Combined>
> {
  /**
   * The portal container which keeps track of all the React Portals containing
   * custom prosemirror NodeViews.
   */
  readonly #portalContainer: PortalContainer = new PortalContainer();

  /**
   * Whether to render the client immediately.
   */
  #getShouldRenderClient: () => boolean | undefined;

  /**
   * Update the should render client state input.
   */
  #setShouldRenderClient: SetShouldRenderClient;

  /**
   * Stores the Prosemirror EditorView dom element.
   */
  #editorRef?: HTMLElement;

  /**
   * Used when suppressHydrationWarning is true to determine when it's okay to
   * render the client content.
   */
  private get shouldRenderClient(): boolean | undefined {
    return this.#getShouldRenderClient();
  }

  /**
   * Keep track of whether the get root props has been called during the most recent render.
   */
  private rootPropsConfig = {
    called: false,
  };

  constructor(parameter: ReactEditorWrapperParameter<Combined>) {
    super(parameter);

    const { getShouldRenderClient, setShouldRenderClient } = parameter;

    this.#getShouldRenderClient = getShouldRenderClient;
    this.#setShouldRenderClient = setShouldRenderClient;

    propIsFunction(this.props.children);

    if (this.manager.view) {
      return;
    }

    this.manager.getPreset(ReactPreset).setOptions({ placeholder: this.props.placeholder ?? '' });
  }

  /**
   * This is called to update props on every render so that values don't become stale.
   */
  update(parameter: ReactEditorWrapperParameter<Combined>) {
    super.update(parameter);

    const { getShouldRenderClient, setShouldRenderClient } = parameter;

    this.#getShouldRenderClient = getShouldRenderClient;
    this.#setShouldRenderClient = setShouldRenderClient;

    return this;
  }

  /**
   * Create the prosemirror editor view.
   */
  protected createView(state: EditorState<SchemaFromCombined<Combined>>) {
    return createEditorView<SchemaFromCombined<Combined>>(
      undefined,
      {
        state,
        nodeViews: this.manager.store.nodeViews,
        dispatchTransaction: this.dispatchTransaction,
        attributes: () => this.getAttributes(),
        editable: () => this.props.editable ?? true,
      },
      this.props.forceEnvironment,
    );
  }

  /**
   * The external `getRootProps` that is used to spread props onto a desired
   * holder element for the prosemirror view.
   */
  private readonly getRootProps = <RefKey extends string = 'ref'>(
    options?: GetRootPropsConfig<RefKey>,
  ) => {
    return this.internalGetRootProps(options, null);
  };

  /**
   * Creates the props that should be spread on the root element inside which
   * the prosemirror instance will be rendered.
   */
  private readonly internalGetRootProps = <RefKey extends string = 'ref'>(
    options?: GetRootPropsConfig<RefKey>,
    children?: ReactNode,
  ): RefKeyRootProps<RefKey> => {
    // Ensure that this is the first time `getRootProps` is being called during
    // this render.
    invariant(!this.rootPropsConfig.called, { code: ErrorConstant.REACT_GET_ROOT_PROPS });
    this.rootPropsConfig.called = true;

    const { refKey: refKey = 'ref', ...config } = options ?? object<NonNullable<typeof options>>();

    return {
      [refKey]: this.onRef,
      key: this.uid,
      ...config,
      children: children ?? this.renderChildren(null),
    } as any;
  };

  /**
   * Stores the Prosemirror editor dom instance for this component using `refs`
   */
  private readonly onRef: Ref<HTMLElement> = (element) => {
    if (element) {
      this.#editorRef = element;
      this.onRefLoad();
    }
  };

  /**
   * Updates the state either by calling onStateChange when it exists or
   * directly setting the internal state via a `setState` call.
   */
  protected updateState(parameter: UpdateStateParameter<SchemaFromCombined<Combined>>) {
    const { state, triggerChange = true, tr, transactions } = parameter;

    if (this.props.value) {
      const { onChange } = this.props;

      invariant(onChange, {
        code: ErrorConstant.REACT_CONTROLLED,
        message:
          'You are required to provide the `onChange` handler when creating a controlled editor.',
      });

      invariant(triggerChange, {
        code: ErrorConstant.REACT_CONTROLLED,
        message:
          'Controlled editors do not support `clearContent` or `setContent` where `triggerChange` is `true`. Update the `value` prop instead.',
      });

      this.onChange({ state, tr });
      return;
    }

    // Update the internal prosemirror state. This happens before we update
    // the component's copy of the state.
    this.view.updateState(state);

    if (triggerChange) {
      this.onChange({ state, tr });
    }

    this.manager.onStateUpdate({ previousState: this.previousState, state, tr, transactions });
  }

  /**
   * Update the controlled state when the value changes and notify the extension
   * of this update.
   */
  updateControlledState(state: EditorState<SchemaFromCombined<Combined>>) {
    this.view.updateState(state);
    this.manager.onStateUpdate({ previousState: this.previousState, state });
  }

  /**
   * Adds the prosemirror view to the dom in the position specified via the
   * component props.
   */
  private addProsemirrorViewToDom(element: HTMLElement, viewDom: Element) {
    if (this.props.insertPosition === 'start') {
      element.insertBefore(viewDom, element.firstChild);
    } else {
      element.append(viewDom);
    }
  }

  /**
   * Called once the container dom node (`this.editorRef`) has been initialized
   * after the component mounts.
   *
   * This method handles the cases where the dom is not focused.
   */
  private onRefLoad() {
    invariant(this.#editorRef, {
      code: ErrorConstant.REACT_EDITOR_VIEW,
      message: 'Something went wrong when initializing the text editor. Please check your setup.',
    });

    const { autoFocus } = this.props;
    this.addProsemirrorViewToDom(this.#editorRef, this.view.dom);

    if (autoFocus) {
      this.focus(autoFocus);
    }

    this.onChange();
    this.addFocusListeners();
  }

  onMount() {
    const { suppressHydrationWarning } = this.props;

    if (suppressHydrationWarning) {
      this.#setShouldRenderClient(true);
    }
  }

  /**
   * Called for every update of the props and state.
   */
  onUpdate(previousEditable: boolean | undefined) {
    // Ensure that `children` is still a render prop
    propIsFunction(this.props.children);

    // Check whether the editable prop has been updated
    if (this.props.editable !== previousEditable && this.view && this.#editorRef) {
      this.view.setProps({ ...this.view.props, editable: () => this.props.editable ?? true });
    }
  }

  get remirrorContext(): RemirrorContextProps<Combined> {
    return {
      ...this.editorWrapperOutput,
      getRootProps: this.getRootProps,
      portalContainer: this.#portalContainer,
    };
  }

  /**
   * Checks whether this is an SSR environment and returns a child array with
   * the SSR component
   *
   * @param children
   */
  private renderChildren(child: ReactNode) {
    const { forceEnvironment, insertPosition = 'end', suppressHydrationWarning } = this.props;

    const children = isArray(child) ? child : [child];

    if (
      shouldUseDomEnvironment(forceEnvironment) &&
      (!suppressHydrationWarning || this.shouldRenderClient)
    ) {
      return children;
    }

    const ssrElement = this.renderSSR();

    return (insertPosition === 'start' ? [ssrElement, ...children] : [...children, ssrElement]).map(
      addKeyToElement,
    );
  }

  /**
   * Return a JSX Element to be used within the domless environment.
   */
  private renderSSR() {
    return (
      <RemirrorSSR
        attributes={this.getAttributes(true)}
        state={this.getState()}
        manager={this.manager}
        editable={this.props.editable ?? true}
      />
    );
  }

  /**
   * Clones the passed element when `getRootProps` hasn't yet been called.
   *
   * This method also supports rendering the children within a domless environment where necessary.
   */
  private renderClonedElement(
    element: JSX.Element,
    rootProperties?: GetRootPropsConfig<string> | boolean,
  ) {
    const { children, ...rest } = getElementProps(element);
    const properties = isPlainObject(rootProperties) ? { ...rootProperties, ...rest } : rest;

    return cloneElement(
      element,
      this.internalGetRootProps(properties, this.renderChildren(children)),
    );
  }

  /**
   * Reset the `getRootProps` called status.
   */
  private prepareRender() {
    // Reset the status of roots props being called
    this.rootPropsConfig.called = false;
  }

  render() {
    this.prepareRender();

    const element: JSX.Element | null = this.props.children({
      ...this.remirrorContext,
    });

    const { children, ...properties } = getElementProps(element);

    if (this.rootPropsConfig.called) {
      // Simply return the element as this method can never actually be called
      // within an ssr environment
      return element;
    } else if (
      // When called by a provider `getRootProps` can't actually be called until
      // the jsx is generated. Check if this is being rendered via any remirror
      // context provider. In this case `getRootProps` **must** be called by the
      // consumer.
      isRemirrorContextProvider(element) ||
      isRemirrorProvider(element)
    ) {
      const { childAsRoot } = element.props;
      return childAsRoot
        ? React.cloneElement(element, properties, this.renderClonedElement(children, childAsRoot))
        : element;
    } else {
      return isReactDOMElement(element) ? (
        this.renderClonedElement(element)
      ) : (
        <div {...this.internalGetRootProps(undefined, this.renderChildren(element))} />
      );
    }
  }
}

interface ReactEditorWrapperParameter<Combined extends AnyCombinedUnion>
  extends EditorWrapperParameter<Combined, ReactEditorProps<Combined>> {
  getShouldRenderClient: () => boolean | undefined;
  setShouldRenderClient: SetShouldRenderClient;
}

type SetShouldRenderClient = Dispatch<SetStateAction<boolean | undefined>>;

/**
 * A hook which creates a reference to the `ReactEditorWrapper` and updates the
 * parameters on every render.
 */
function useEditorWrapper<Combined extends AnyCombinedUnion>(
  parameter: ReactEditorWrapperParameter<Combined>,
) {
  return useRef(new ReactEditorWrapper<Combined>(parameter)).current.update(parameter);
}

/**
 * If no string handler is provided, but the user tries to provide a string as
 * content then throw an error.
 */
function defaultStringHandler(): never {
  invariant(false, {
    code: ErrorConstant.REACT_EDITOR_VIEW,
    message:
      'No valid string handler. In order to pass in `string` as `initialContent` to the remirror editor you must provide a valid stringHandler prop',
  });
}
