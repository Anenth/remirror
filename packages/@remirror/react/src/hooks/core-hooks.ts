import isEqual from 'fast-deep-equal/react';
import { DependencyList, useEffect, useLayoutEffect, useRef, useState } from 'react';
import useEffectOnce from 'react-use/lib/useEffectOnce';
import useSetState from 'react-use/lib/useSetState';
import useUpdateEffect from 'react-use/lib/useUpdateEffect';
import ResizeObserver from 'resize-observer-polyfill';

/**
 * Preserves the previous version of a provided value.
 *
 * ```tsx
 * const [isOpen, setOpen] = useState<boolean>(false)
 * const previous = usePrevious(isOpen)
 *
 * return <span onClick={() => setOpen(!isOpen)}>{isOpen && previous === isOpen ? 'Stable' : 'Unstable' }</span>
 * ```
 */
export function usePrevious<Value>(value: Value) {
  const ref = useRef<Value>();
  useEffect(() => void (ref.current = value), [value]);
  return ref.current;
}

export interface DOMRectReadOnlyLike {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
  readonly left: number;
}

const defaultBounds = { x: 0, y: 0, width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 };

/**
 * Provides the measurements for a react element at the point of layout.
 *
 * @remarks
 *
 * Taken from https://codesandbox.io/embed/lp80n9z7v9
 *
 * ```tsx
 * const [bindRef, { height }] = useMeasure()
 *
 * return <div {...bindRef}>Height: {height}</div>
 * ```
 */
export function useMeasure<Ref extends HTMLElement = any>() {
  const ref = useRef<Ref>(null);
  const [bounds, setBounds] = useState<DOMRectReadOnlyLike>(defaultBounds);

  useLayoutEffect(() => {
    const resizeObserver = new ResizeObserver(([entry]) => setBounds(entry.contentRect));

    if (ref.current) {
      resizeObserver.observe(ref.current);
    }

    return () => resizeObserver.disconnect();
  }, []);

  return [{ ref }, bounds] as const;
}

/**
 * The `setState` type for the `useSetState` hook.
 */
export type PartialDispatch<Type extends object> = (
  patch: Partial<Type> | ((prevState: Type) => Partial<Type>),
) => void;

/**
 * A `useEffect` function with a warning the provided dependencies are deeply
 * equal.
 */
const useEffectWithWarning: typeof useEffect =
  process.env.NODE_ENV === 'production'
    ? useEffect
    : (effect, deps) => {
        const ref = useRef<DependencyList>();
        const unnecessaryChange = useRef(0);

        useEffectOnce(() => {
          ref.current = deps;
        });

        useUpdateEffect(() => {
          if (!isEqual(deps, ref.current)) {
            unnecessaryChange.current = 0;
            ref.current = deps;
            return;
          }

          unnecessaryChange.current += 1;
        });

        const wrappedEffect = () => {
          if (unnecessaryChange.current >= 1) {
            console.warn(
              `The dependencies passed into your useEffect are deeply equal, but an update has been triggered ${unnecessaryChange.current} time(s). Please consider \`useMemo\` to memoize your dependencies to prevent unwanted re-renders.`,
              deps,
            );
          }

          return effect();
        };

        // eslint-disable-next-line react-hooks/exhaustive-deps
        useEffect(wrappedEffect, deps);
      };

export { useEffectWithWarning, useSetState };
