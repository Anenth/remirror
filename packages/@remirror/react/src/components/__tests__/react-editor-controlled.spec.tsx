import { act, render } from '@testing-library/react';
import { RemirrorTestChain } from 'jest-remirror';
import React, { FC, useState } from 'react';

import {
  AnyCombinedUnion,
  EditorState,
  fromHtml,
  PlainExtension,
  SchemaFromCombined,
  StateUpdateLifecycleMethod,
} from '@remirror/core';

import { createReactManager } from '../../hooks/editor-hooks';
import { RemirrorContextProps } from '../../react-types';
import { ReactEditor } from '../react-editor';

const label = 'Remirror editor';

function create<Combined extends AnyCombinedUnion>(combined?: Combined[]) {
  const manager = createReactManager(combined ?? []);
  const chain = RemirrorTestChain.create(manager);

  return {
    manager,
    chain,
    doc: chain.nodes.doc,
    p: chain.nodes.paragraph,
    props: {
      label,
      stringHandler: fromHtml,
    },
  };
}

describe('Remirror Controlled Component', () => {
  it('should set the initial value', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>This is the initial value</p>',
      stringHandler: fromHtml,
    });
    const onChange = jest.fn();

    const { getByRole } = render(
      <ReactEditor {...props} value={value} manager={manager} onChange={onChange}>
        {() => <div />}
      </ReactEditor>,
    );

    expect(getByRole('textbox')).toMatchSnapshot();

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0].firstRender).toBeTrue();
  });

  it('overrides initial content', () => {
    const { manager, props, chain } = create();

    const value = manager.createState({
      content: '<p>Not terrible</p>',
      stringHandler: fromHtml,
    });
    const onChange = jest.fn();

    render(
      <ReactEditor
        {...props}
        manager={manager}
        initialContent='<p>Terrible</p>'
        value={value}
        onChange={onChange}
      >
        {() => <div />}
      </ReactEditor>,
    );

    expect(chain.view.dom).toMatchSnapshot();
  });

  it('responds to updates to the editor state', () => {
    const { manager, props, chain } = create();

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: '<p>some content</p>',
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            if (parameter.firstRender) {
              return;
            }

            setValue(parameter.state);
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    const { rerender } = render(<Component />);

    act(() => {
      chain.dispatchCommand(({ dispatch, tr }) => dispatch(tr.insertText('add more value ')));
    });

    rerender(<Component />);

    expect(chain.view.dom).toMatchSnapshot();
  });

  it('can override updates to the editor state', () => {
    const { manager, props, chain } = create();

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: '<p>some content</p>',
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            const { firstRender, createStateFromContent, getText } = parameter;

            if (firstRender) {
              return;
            }

            setValue(createStateFromContent(`<p>Hello</p><p>${getText()}</p>`));
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    const { rerender } = render(<Component />);

    act(() => {
      chain.dispatchCommand(({ dispatch, tr }) => dispatch(tr.insertText('add more value ')));
    });

    rerender(<Component />);

    expect(chain.view.dom).toMatchSnapshot();
  });

  it('throws when using  `setContent` updates', () => {
    const { manager, props } = create();

    const value = manager.createState({
      content: '<p>some content</p>',
      stringHandler: fromHtml,
    });

    let ctx: RemirrorContextProps<typeof manager['~EP']>;

    const Component: FC<{
      editorState: EditorState<SchemaFromCombined<typeof manager['~EP']>>;
    }> = ({ editorState }) => {
      return (
        <ReactEditor {...props} value={editorState} manager={manager} onChange={jest.fn()}>
          {(context) => {
            ctx = context as any;
            return <div />;
          }}
        </ReactEditor>
      );
    };

    render(<Component editorState={value} />);

    expect(() => ctx.setContent('<p>Error</p>')).toThrowErrorMatchingSnapshot();
    expect(() => ctx.clearContent()).toThrowErrorMatchingSnapshot();
  });

  it('notifies extensions of state updates via `manager.onStateUpdate`', () => {
    const mock = jest.fn();

    class UpdateExtension extends PlainExtension {
      get name() {
        return 'update' as const;
      }

      onStateUpdate: StateUpdateLifecycleMethod = mock;
    }

    const { manager, props, chain, doc, p } = create([new UpdateExtension()]);

    const Component = () => {
      const [value, setValue] = useState<EditorState>(
        manager.createState({
          content: doc(p('some content')),
          stringHandler: fromHtml,
        }),
      );

      return (
        <ReactEditor
          {...props}
          value={value}
          manager={manager}
          onChange={(parameter) => {
            const { firstRender, state } = parameter;

            if (firstRender) {
              return;
            }

            setValue(state);
          }}
        >
          {() => <div />}
        </ReactEditor>
      );
    };

    render(<Component />);

    act(() => {
      chain.commands.insertText('First text update');
    });

    expect(mock).toHaveBeenCalledTimes(2);
    expect(mock.mock.calls[1][0].state).toBe(chain.state);
  });
});
