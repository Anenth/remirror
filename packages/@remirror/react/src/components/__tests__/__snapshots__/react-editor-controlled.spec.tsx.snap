// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remirror Controlled Component can override updates to the editor state 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  autofocus="false"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    Hello
  </p>
  <p>
    add more value some content
  </p>
</div>
`;

exports[`Remirror Controlled Component overrides initial content 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  autofocus="false"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    Not terrible
  </p>
</div>
`;

exports[`Remirror Controlled Component responds to updates to the editor state 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  autofocus="false"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    add more value some content
  </p>
</div>
`;

exports[`Remirror Controlled Component should set the initial value 1`] = `
<div
  aria-label="Remirror editor"
  aria-multiline="true"
  aria-placeholder=""
  autofocus="false"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
>
  <p>
    This is the initial value
  </p>
</div>
`;

exports[`Remirror Controlled Component throws when using  \`setContent\` updates 1`] = `
"There is a problem with your controlled editor setup.

Controlled editors do not support \`clearContent\` or \`setContent\` where \`triggerChange\` is \`true\`. Update the \`value\` prop instead.

For more information visit https://docs.remirror.org/errors#RMR0202"
`;

exports[`Remirror Controlled Component throws when using  \`setContent\` updates 2`] = `
"There is a problem with your controlled editor setup.

Controlled editors do not support \`clearContent\` or \`setContent\` where \`triggerChange\` is \`true\`. Update the \`value\` prop instead.

For more information visit https://docs.remirror.org/errors#RMR0202"
`;
